File: CSharp/OpenTraceability/Utility/Attributes/OpenTraceabilityProductsAttribute.cs
using OpenTraceability.Models.Events;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OpenTraceability.Utility.Attributes
{
    public enum OpenTraceabilityProductsListType
    {
        EPCList,
        QuantityList
    }

    [AttributeUsage(AttributeTargets.Property, AllowMultiple = true)]
    public class OpenTraceabilityProductsAttribute : Attribute
    {
        public string Name { get; set; }
        public EPCISVersion? Version { get; set; }
        public EventProductType ProductType { get; set; }
        public int? SequenceOrder { get; set; }
        public OpenTraceabilityProductsListType ListType { get; set; }
        public bool Required { get; set; } = false;

        public OpenTraceabilityProductsAttribute(string name, EPCISVersion version, EventProductType productType, int SequenceOrder, OpenTraceabilityProductsListType listType)
        {
            this.Name = name;
            this.Version = version;
            this.ProductType = productType;
            this.SequenceOrder = SequenceOrder;
            this.ListType = listType;
        }

        public OpenTraceabilityProductsAttribute(string name, EventProductType productType, int SequenceOrder, OpenTraceabilityProductsListType listType)
        {
            this.Name = name;
            this.ProductType = productType;
            this.SequenceOrder = SequenceOrder;
            this.ListType = listType;
        }
    }
}

File: JDK11/opentraceability/src/main/java/opentraceability/utility/attributes/OpenTraceabilityProductsAttribute.java
package opentraceability.utility.attributes;

import opentraceability.models.events.EPCISVersion;
import opentraceability.models.events.EventProductType;

import java.lang.annotation.ElementType;
import java.lang.annotation.Repeatable;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Repeatable(OpenTraceabilityProductsAttributes.class)
@Retention(RetentionPolicy.RUNTIME)
public @interface OpenTraceabilityProductsAttribute {
    public String name();
    public EPCISVersion version();
    public EventProductType productType();
    public OpenTraceabilityProductsListType listType();
    public boolean required() default false;
}

